File: solutions/vbnet/grains/1/Grains.vb
Imports System

Public Module Grains
    Public Function Square(ByVal n As Integer) As ULong
        If n <= 0 Then
            Throw New ArgumentOutOfRangeException()
        End If
        If n > 64 Then
            Throw New ArgumentOutOfRangeException()
        End If
        Return 1UL << (n - 1)
    End Function
    
    Public Function Total() As ULong
        Return ULong.MaxValue
    End Function
End Module

File: solutions/delphi/grains/1/grains.pas
unit UGrains;

interface

type
  Grains = class
  public
    class function Square(const Avalue:Integer):UInt64; static;
    class function Total:UInt64; static;
  end;

const
  MIN_VALUE = 1;
  MAX_VALUE = 64;

implementation

uses
  System.SysUtils;

class function Grains.Square(const AValue: Integer): UInt64;
begin
  if (AValue < 1) or (Avalue > MAX_VALUE) then
    raise ERangeError.Create('El valor debe estar entre 1..64')
  else if (AValue = 1) then
    Result := 1
  else
    Result := Grains.Square(AValue-1)*2;
end;



class function Grains.Total: UInt64;
begin
  Result := (Grains.Square(MAX_VALUE) * 2) - 1;
end;

end.